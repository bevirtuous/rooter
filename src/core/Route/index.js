import { parseUrl } from 'query-string';

const parseOptions = {
  arrayFormat: 'comma',
  parseBooleans: true,
  parseNumbers: true,
};

function Route(options) {
  const { meta = {} } = options;
  const location = options.pathname.trim();
  const id = options.id;

  const splitPath = location.split('#');
  const path = splitPath[0];
  const hash = splitPath[1] || null;
  const { query, url: pathname } = parseUrl(path, parseOptions);

  return {
    id,
    hash,
    location,
    meta,
    pathname,
    query,
  };
}

export default Route;
