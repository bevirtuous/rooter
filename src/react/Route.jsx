import React from 'react';
import Context from './context';
import usePath from './usePath';
import useLocation from './useLocation';

function Route({ children, exact = false, path }) {
  const location = useLocation();
  const { isExactMatch, isMatch, params } = usePath(path);
  const isVisible = exact ? isExactMatch : isMatch;

  if (!isVisible) {
    return null;
  }

  return (
    <Context.Provider value={{ ...location, params, pattern: path }}>
      {children}
    </Context.Provider>
  );
}

export default Route;