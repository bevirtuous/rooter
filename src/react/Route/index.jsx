import React from 'react';
import PropTypes from 'prop-types';
import router from '../../core/Router';
import RouteNotFound from '../404';
import { RouteContext, RouterContext } from '../context';

class Route extends React.Component {
  constructor(props) {
    super(props);
    router.register(props.path);
  }

  render() {
    const { children, path } = this.props;
    const route = router.getCurrentRoute();

    if (route.pattern !== path) {
      return null;
    }

    const key = `${route.id}-${route.pathname}`;

    return (
      <RouteContext.Provider key={key} value={route}>
        {children}
      </RouteContext.Provider>
    );
  }
}

Route.contextType = RouterContext;
Route.NotFound = RouteNotFound;
Route.propTypes = {
  children: PropTypes.node.isRequired,
  path: PropTypes.string.isRequired,
};

export default Route;
