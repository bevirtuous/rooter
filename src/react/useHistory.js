import history from 'history/browser';

function useHistory() {
  return {
    back: history.back,
    push: history.push,
  };
}

export default useHistory;
