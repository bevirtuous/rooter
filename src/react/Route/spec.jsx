import React from 'react';
import { act } from 'react-dom/test-utils';
import { mount } from 'enzyme';
import router from '../../core/Router';
import Router from '../Router/index';
import { RouteContext } from '../context';
import Route from './index';

const path = '/myroute/:id';
const spy = jest.spyOn(router, 'register');

describe('<Route />', () => {
  it('should render component', () => {
    const app = mount((
      <Router>
        <Route path={path}>
          <div />
        </Route>
      </Router>
    ));

    expect(spy).toHaveBeenCalledWith(path);
    expect(app).toMatchSnapshot();
  });

  it('should correctly set the RouteContext value', () => {
    let contextValue = null;
    const current = router.getCurrentRoute();

    const MyComponent = () => (
      <RouteContext.Consumer>
        {(route) => {
          contextValue = route;
          return null;
        }}
      </RouteContext.Consumer>
    );

    mount((
      <Router>
        <Route path={path}>
          <MyComponent />
        </Route>
      </Router>
    ));

    expect(current).toMatchObject(contextValue);
  });

  it('should render null when does not match current route', () => {
    const route = mount((
      <Router>
        <Route path="/wrong">
          <div />
        </Route>
      </Router>
    ));

    expect(route.html()).toBe('');
  });

  it('should react to router events and update', async () => {
    const route = mount((
      <Router>
        <Route path={path}>
          <div />
        </Route>
        <Route path="/other">
          <div />
        </Route>
      </Router>
    ));

    // Should initially render the component (match found).
    expect(route.find(Route).at(0).html()).toBe('<div></div>');
    expect(route.find(Route).at(1).html()).toBeNull();

    await act(async () => {
      await router.push({ to: '/other' });
    });

    route.update();

    // Should now render null (first route does not match).
    expect(route.find(Route).at(0).html()).toBeNull();
    expect(route.find(Route).at(1).html()).toBe('<div></div>');
  });
});
