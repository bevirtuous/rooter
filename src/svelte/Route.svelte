<script>
  import { register } from '../core';
  import store from './store';

  export let component;
  export let path;

  register(path);
</script>

{#if $store.current && path === $store.current.pattern}
  <svelte:component this={component} />
{/if}
