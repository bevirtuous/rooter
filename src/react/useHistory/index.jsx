import { useContext } from 'react';
import router from '../../core/Router';
import stack from '../../core/Stack';
import { RouterContext } from '../context';

const useHistory = () => {
  const { next, prev } = useContext(RouterContext);

  return {
    current: stack.get(next),
    currentIndex: router.currentIndex,
    previous: stack.get(prev),
  };
};

export default useHistory;
