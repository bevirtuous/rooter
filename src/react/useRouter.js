import { useEffect } from 'react';
import history from 'history/browser';
import useStore from './store';

function useRouter(plugins = []) {
  const back = useStore(state => state.back);
  const push = useStore(state => state.push);

  useEffect(() => {
    const listener = history.listen(({ action, location }) => {
      if (action === 'POP') {
        back();
      } else {
        push(location);
      }
    });

    return () => listener();
  }, []);

  return 123;
}

export default useRouter;
