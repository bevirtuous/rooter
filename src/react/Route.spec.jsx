import React from 'react';
import { renderHook } from '@testing-library/react-hooks';
import { act } from 'react-dom/test-utils';
import { mount, shallow } from 'enzyme';
import Route from './Route';
import useHistory from './useHistory';
import useRouter from './useRouter';

describe('React - <Route />', () => {
  it('should render when matched to history', () => {
    const app = shallow((
      <Route path="/myroute/:id">
        <div>Hello World.</div>
      </Route>
    ));

    expect(app).toMatchSnapshot();
  });

  it('should not render when not matched to history', () => {
    const app = shallow((
      <Route path="/myroute/:id/further">
        <div>Hello World.</div>
      </Route>
    ));

    expect(app.html()).toBeNull();
  });

  it('should only render when exact match is found', () => {
    const app = shallow((
      <Route path="/myroute" exact>
        <div>Hello World.</div>
      </Route>
    ));

    expect(app.html()).toBeNull();
  });

  it('should only render index when is exact match', () => {
    const app = shallow((
      <Route path="/">
        <div>Hello World.</div>
      </Route>
    ));

    expect(app.html()).toBeNull();
  });

  it('should react to history events', (done) => {
    renderHook(() => useRouter());

    const history = useHistory();
    const app = mount((
      <div>
        <Route path="/myroute/:id">
          <div>First route</div>
        </Route>
        <Route path="/test">
          <div>Second route</div>
        </Route>
      </div>
    ));

    expect(app.find(Route).at(0).html()).toBe('<div>First route</div>');
    expect(app.find(Route).at(1).html()).toBeNull();

    act(() => {
      history.push('/test');
    });

    setTimeout(() => {
      app.update();
      expect(app.find(Route).at(0).html()).toBeNull();
      expect(app.find(Route).at(1).html()).toBe('<div>Second route</div>');
      done();
    }, 1000)
  });
});
