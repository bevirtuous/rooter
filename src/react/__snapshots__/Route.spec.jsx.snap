// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React - <Route /> should render when matched to history 1`] = `
<ContextProvider
  value={
    Object {
      "fragment": "",
      "id": "default",
      "params": Object {
        "id": "123",
      },
      "pathname": "/myroute/123",
      "pattern": "/myroute/:id",
      "search": "",
      "state": Object {},
    }
  }
>
  <div>
    Hello World.
  </div>
</ContextProvider>
`;
